Add helper to rewrite default-env.txt files

bjorncs opened the pull request.

bjorncs committed:
Add helper to rewrite default-env.txt files

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/env/DefaultEnvRewriter.java:
```diff
@@ -0,0 +1,101 @@
+        return this;
+    }
+
+    public boolean converge() throws IOException {
```
There is no need to expose IOException to the caller? Below you can use `uncheck(() -> Files.write())`, or use UnixPath instead of Path.

bjorncs replied:
Done

hakonhall commented on node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/util/env/DefaultEnvRewriterTest.java:
```diff
@@ -0,0 +1,48 @@
+
+        DefaultEnvRewriter rewriter = new DefaultEnvRewriter(tempFile);
+        rewriter.addOverride("VESPA_HOSTNAME", "my-new-hostname");
+        rewriter.addFallback("VESPA_CONFIGSERVER", "new-fallback-configserver");
```
can values be quoted or contain space?

bjorncs replied:
Both are accepted based on my interpretation of the Perl code parsing the file:
https://github.com/vespa-engine/vespa/blob/master/vespabase/src/common-env.sh#L35

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/env/package-info.java:
```diff
@@ -0,0 +1,8 @@
+ * @author bjorncs
+ */
+@ExportPackage
+package com.yahoo.vespa.hosted.node.admin.task.util.env;
```
What other files do you expect in this package? For now 1 file in util has been traded with 1 directory in util, and 2 files in util/env.

bjorncs replied:
Currently no other files. I am fine with moving it to the parent package.

bjorncs replied:
Done

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/env/DefaultEnvRewriter.java:
```diff
@@ -0,0 +1,101 @@
+        List<String> newDefaultEnvLines = new ArrayList<>();
+        Set<String> seenNames = new TreeSet<>();
+        for (String line : defaultEnvLines) {
+            String[] items = line.split(" ");
```
As discussed offline, the value (starting with the 3. token) 1. may be quoted, and 2. may contain space.

bjorncs committed:
Misc changes

- Move to parent package
- Use Exceptions.uncheck()
- Write changes atomically

freva commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/DefaultEnvRewriter.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public boolean converge() {
+        List<String> defaultEnvLines = uncheck(() -> Files.readAllLines(defaultEnvFile));
```
Is the file always guaranteed to exist? Maybe better with
```suggestion
        List<String> defaultEnvLines = ifExists(() -> Files.readAllLines(defaultEnvFile)).orElseGet(List::of);
```

bjorncs replied:
It's part of the Vespa installation. I'll change the code to handle missing file anyway.

freva reviewed (approved):
We also write default env for the container in `host-admin`, would be nice if we could generalize this class slightly so I can be reused there:
1. Create empty constructor and instead take in the `defaultEnvFile` as argument to `converge()`.
2. Add method to get value as `String` independent of pre-existing `defaultEnvFile`.

bjorncs committed:
Handle non-existing default-env.txt file

bjorncs committed:
Add method to generate default-env.txt content to string

- Change name of converge method to updateFile
- Remove default-env.txt path from constructor, add as parameter to updateFile instead.

bjorncs commented:
@freva PTAL

bjorncs committed:
Rename DefaultEnvRewriter -> DefaultEnvWriter

freva reviewed (approved):
:100:
